252 input validation

najimgh opened the pull request:
Input validation added for the AddAsset form, AddAssetType form and the AssetTypeInfo form

resolves #252 and #198

najimgh committed:
Issue (#252)
- Added input validation on the add system form
- Added red border when input is wrong

najimgh committed:
Merge branch 'develop' of https://github.com/rdsaliba/notorious-eng into 252_input_validation

 Conflicts:
	JavaFXUserInterface/src/main/java/Controllers/AddAssetController.java
	JavaFXUserInterface/src/main/resources/AddAsset.fxml

najimgh committed:
Issue (#252)
- Added input validation functionality to the AddAssetType form, AddAsset form and AssetTypeInfo form
- Refactored constants into the TextConstants file
- Refactored methods into the UIUtilities file

JayChouteau commented on JavaFXUserInterface/src/main/java/Controllers/AddAssetController.java:
```diff
@@ -27,8 +29,8 @@
-    private static final String ERROR_DIALOG = "Error Dialog";
-    private static final String ERROR_HEADER = "Please enter values for all text fields.";
+//    private static final String ERROR_DIALOG = "Error Dialog";
+//    private static final String ERROR_HEADER = "Please enter values for all text fields.";
```
Why is there commented-out code? This should be either deleted or changed.

najimgh replied:
Removed

JayChouteau commented on JavaFXUserInterface/src/main/java/Controllers/AddAssetController.java:
```diff
@@ -173,26 +180,103 @@ void saveDialog(MouseEvent mouseEvent) {
+//
+//        alert.showAndWait();
+//    }
+
```
Same thing as previous. Commented-out code should be removed or changed.

najimgh replied:
Removed

JayChouteau commented on JavaFXUserInterface/src/main/java/Utilities/UIUtilities.java:
```diff
@@ -166,6 +165,115 @@ public void changeScene(MouseEvent mouseEvent, String fxmlFileName, Asset asset)
+            errorMessages[i] = null;
+        }
+    }
+
```
There is a lot of duplication due to the repetition of removeInputError() and createInputError(). The only difference is if the passing argument is a textArea or a textField. It would be better to remove the duplication by having 1 method for each action that allows the passing of an element (either textarea or textfield) and then have a branch condition depending on which one it is within the method. I believe both textarea and textfield are children of TextComponent. This means you could have a TextComponent as a parameter, allowing the function to take in either a textArea or textField and making the distinction within the function.

najimgh replied:
Resolved. The function takes a TextInputControl instead of a textArea or TextField

JayChouteau replied:
Nice job

JayChouteau commented on JavaFXUserInterface/src/main/resources/css/addAsset.css:
```diff
@@ -4,4 +4,15 @@
+.input-error {
+    -fx-border-color: red;
+    -fx-border-width: 2px;
 }
```
There is a lot of duplication between the CSS files for the same code. it would be better to put this on the same CSS file and have it named something more general. Maybe you can reference multiple CSS files in a FXML as well. There is too much duplication.

najimgh replied:
Moved CSS styles to the style.css file

JayChouteau commented on JavaFXUserInterface/src/main/java/Controllers/AddAssetTypeController.java:
```diff
@@ -135,4 +146,89 @@ public boolean saveAssetType(AssetType assetType) {
+                UIUtilities.removeInputError(inputError, errorMessages, validInput, thresholdFailedValue, 6);
+            }
+        }
+        return validForm;
```
There is no checking for all the above previous thresholds compared to a threshold, it only compares to the first previous. This means that there are no validation if one value is missing. For example, if your warning threshold has no value, then the failed value can be set higher or lower than the caution value or the ok since the validation on failed is made only compared to warning. It would be important to do the validation compared to the first previous non-empty valid input and not just the first previous.

najimgh replied:
This has been resolved. I have also refactored the code a little more, removing some duplicated code

JayChouteau commented on JavaFXUserInterface/src/main/java/Controllers/AddAssetController.java:
```diff
@@ -173,26 +180,103 @@ void saveDialog(MouseEvent mouseEvent) {
+        if (assetNameValue.trim().isEmpty() || assetNameValue.length() > 50) {
+            validForm = false;
+            validInput[0] = false;
+            UIUtilities.createInputError(inputError, errorMessages, assetNameInput, TextConstants.EMPTY_FIELD_ERROR + " and/or \n" + TextConstants.MAX_50_CHARACTERS_ERROR, 65.0, horizontalPosition, 0);
```
This error message is confusing. It would be better to have those 2 conditions separated and have the corresponding message printed depending on the condition triggered.

najimgh replied:
Resolved

JayChouteau commented on JavaFXUserInterface/src/main/java/Utilities/TextConstants.java:
```diff
@@ -20,6 +20,17 @@
+    public static final String LETTER_NUMBER_ERROR = "Value can only contain letters or numbers";
+    public static final String ADIVSORY_CAUTION = "The Advisory Threshold needs to be larger \nthan the Caution Threshold";
+    public static final String CAUTION_WARNING = "The Caution Threshold needs to be larger \nthan the Warning Threshold";
+    public static final String WARNING_FAILED = "The Warning Threshold needs to be larger \nthan the Failed Threshold";
```
This relates to my previous comment about doing validation on just the previous threshold instead of doing validation on the previous valid threshold. Most likely, the message would be changed.

najimgh replied:
New error messages added

JayChouteau replied:
Well done.

JayChouteau reviewed (changes requested):
The way the input validation errors are highlighted to the user is well done: having the error text fields highlighted in red is clear. However, there are some error messages that are a bit confusing (see comments in the review). All input validations as described in the issue have been performed and work as intended. However, there is one small issue with threshold input validation (see review comments). My last comment is about code duplication. There is a lot of code duplication in this pull-request and it would be better to find ways to get rid of it.

najimgh committed:
Issue (#252)
- Removed code duplicates
- Removed css style duplicates
- Fixed the error message being displayed for the asset name
- Removed commented code
- Fixed issue #198
   - Serial number can now contain letters, numbers or hyphens

najimgh committed:
Issue (#252)
- Thresholds are now compare to all Thresholds above if they contain values
- Refactoring duplicated code

najimgh committed:
Merge branch 'develop' into 252_input_validation

JayChouteau reviewed (approved):
Thanks for addressing the comments. This pull-request is approved now.
